File: SmartReceipts/Reports/Generators/TripFullPDFGenerator.m
//
//  TripFullPDFGenerator.m
//  SmartReceipts
//
//  Created by Jaanus Siim on 25/04/15.
//  Copyright (c) 2015 Will Baumann. All rights reserved.
//

#import "TripFullPDFGenerator.h"
#import "WBPdfDrawer.h"
#import "WBTrip.h"
#import "WBDateFormatter.h"
#import "ReportPDFTable.h"
#import "WBPreferences.h"
#import "Database+Distances.h"
#import "WBReceipt.h"
#import "DistancesToReceiptsConverter.h"
#import "NSString+Validation.h"
#import "Distance.h"
#import "Constants.h"
#import "PrettyPDFRender.h"
#import "SmartReceipts-Swift.h"

@implementation TripFullPDFGenerator

- (BOOL)generateToPath:(NSString *)outputPath {
    // render tables in landscape orientation if needed
    self.pdfRender.landscapePreferred = [WBPreferences printReceiptTableLandscape];
    
    if (![self.pdfRender setOutputPath:outputPath]) {
        return NO; // error
    }
    
    [self appendSummaryAndTables];
    
    // always render fullPageElements and images in protrait orientation
    self.pdfRender.landscapePreferred = NO;
    [self.pdfRender startNextPage];
    [self appendImages];

    BOOL success = [self.pdfRender renderPages];
    return success;
}

- (void)appendSummaryAndTables {
    PricesCollection *netTotal = [[PricesCollection alloc] init];
    PricesCollection *receiptTotal = [[PricesCollection alloc] init];
    PricesCollection *reimbursableTotal = [[PricesCollection alloc] init];
    PricesCollection *noTaxesTotal = [[PricesCollection alloc] init];
    PricesCollection *taxesTotal = [[PricesCollection alloc] init];
    PricesCollection *distanceTotal = [[PricesCollection alloc] init];

    NSArray *receipts = [self receipts];

    BOOL pricesPreTax = [WBPreferences enteredPricePreTax];
    BOOL reportOnlyReimbursable = [WBPreferences onlyIncludeReimbursableReceiptsInReports];

    for (WBReceipt *receipt in receipts) {
        if (reportOnlyReimbursable && !receipt.isReimbursable) {
            continue;
        }

        [netTotal addPrice:receipt.targetPrice];
        [receiptTotal addPrice:receipt.targetPrice];
        [noTaxesTotal addPrice:receipt.targetPrice];
        [noTaxesTotal subtractPrice:receipt.targetTax];
        [taxesTotal addPrice:receipt.targetTax];
        if (pricesPreTax) {
            [netTotal addPrice:receipt.targetTax];
        }
        if (receipt.isReimbursable) {
            [reimbursableTotal addPrice:receipt.targetPrice];
        }
    }

    NSArray *distances = [self distances];

    for (Distance *distance in distances) {
        [netTotal addPrice:distance.totalRate];
        [distanceTotal addPrice:distance.totalRate];
    }

    [self.pdfRender setTripName:self.trip.name];

    if (![receiptTotal isEqual:netTotal]) {
        [self.pdfRender appendHeaderRow:[NSString stringWithFormat:@"%@ %@", NSLocalizedString(@"pdf.report.receipts.total.label", nil), receiptTotal.currencyFormattedPrice]];
    }

    if ([WBPreferences includeTaxField]) {
        if (pricesPreTax && taxesTotal.hasValue) {
            [self.pdfRender appendHeaderRow:[NSString stringWithFormat:@"%@ %@", NSLocalizedString(@"pdf.report.tax.total.label", nil), taxesTotal.currencyFormattedPrice]];
        } else if (![noTaxesTotal isEqual:receiptTotal] && noTaxesTotal.hasValue) {
            [self.pdfRender appendHeaderRow:[NSString stringWithFormat:@"%@ %@", NSLocalizedString(@"pdf.report.receipts.total.sans.tax.label", nil), noTaxesTotal.currencyFormattedPrice]];
        }
    }

    if (!reportOnlyReimbursable && ![reimbursableTotal isEqual:receiptTotal]) {
        [self.pdfRender appendHeaderRow:[NSString stringWithFormat:@"%@ %@", NSLocalizedString(@"pdf.report.receipts.total.reimbursable.label", nil), reimbursableTotal.currencyFormattedPrice]];
    }

    if (distances.count > 0) {
        [self.pdfRender appendHeaderRow:[NSString stringWithFormat:@"%@ %@", NSLocalizedString(@"pdf.report.distance.total.label", nil), distanceTotal.currencyFormattedPrice]];
    }

    [self.pdfRender appendHeaderRow:[NSString stringWithFormat:@"%@ %@", NSLocalizedString(@"pdf.report.gross.total.label", nil), netTotal.currencyFormattedPrice]];


    [self.pdfRender appendHeaderRow:[NSString stringWithFormat:NSLocalizedString(@"pdf.report.from.to.label.base", nil),
                                                           [self.dateFormatter formattedDate:[self.trip startDate] inTimeZone:[self.trip startTimeZone]],
                                                           [self.dateFormatter formattedDate:[self.trip endDate] inTimeZone:[self.trip endTimeZone]]
    ]];

    if ([WBPreferences trackCostCenter] && self.trip.costCenter.hasValue) {
        [self.pdfRender appendHeaderRow:[NSString stringWithFormat:@"%@ %@", NSLocalizedString(@"pdf.report.const.center.label", nil), self.trip.costCenter]];
    }

    if (self.trip.comment.hasValue) {
        [self.pdfRender appendHeaderRow:[NSString stringWithFormat:@"%@ %@", NSLocalizedString(@"pdf.report.comment.label", nil), self.trip.comment]];
    }

    if (distances.count > 0) {
        [self.pdfRender appendHeaderRow:[NSString stringWithFormat:NSLocalizedString(@"pdf.report.distance.traveled.label.base", nil), [self.database totalDistanceTraveledForTrip:self.trip].floatValue]];
    }

    [self.pdfRender closeHeader];

    ReportPDFTable *receiptsTable = [[ReportPDFTable alloc] initWithPDFRender:self.pdfRender columns:[self receiptColumns]];
    [receiptsTable setIncludeHeaders:YES];
    if ([WBPreferences printDailyDistanceValues]) {
        NSArray *distanceReceipts = [DistancesToReceiptsConverter convertDistances:distances];
        receipts = [receipts arrayByAddingObjectsFromArray:distanceReceipts];
        receipts = [receipts sortedArrayUsingComparator:^NSComparisonResult(id obj1, id obj2) {
            WBReceipt *one = obj1;
            WBReceipt *two = obj2;
            return [two.date compare:one.date];
        }];
    }

    [receiptsTable appendTableWithRows:receipts];

    if (![WBPreferences printDistanceTable]) {
        return;
    }

    if (distances.count == 0) {
        return;
    }

    ReportPDFTable *distancesTable = [[ReportPDFTable alloc] initWithPDFRender:self.pdfRender columns:[self distanceColumns]];
    [distancesTable setIncludeHeaders:YES];
    [distancesTable appendTableWithRows:distances];
}

@end

File: SmartReceipts/Reports/Generators/TripFullPDFGenerator.swift
//
//  TripFullPDFGenerator.swift
//  SmartReceipts
//
//  Created by Bogdan Evsenev on 03/12/2017.
//  Copyright © 2017 Will Baumann. All rights reserved.
//

import Foundation

class TripFullPDFGenerator: TripImagesPDFGenerator {
    override func generateTo(path: String) -> Bool {
        // render tables in landscape orientation if needed
        pdfRender.portraitPreferred = !WBPreferences.printReceiptTableLandscape()
        
        if !pdfRender.setOutput(path: path) { return false }
        appendSummaryAndTables()

        // always render fullPageElements and images in protrait orientation
        pdfRender.portraitPreferred = true
        pdfRender.startNextPage()
        appendImages()
        return pdfRender.renderPages()
    }
    
    private func appendSummaryAndTables() {
        let receiptTotal = PricesCollection()
        let reimbursableTotal = PricesCollection()
        let noTaxesTotal = PricesCollection()
        let taxesTotal = PricesCollection()
        let distanceTotal = PricesCollection()
        let grandTotal = PricesCollection()
        let grandTotalReimbursable = PricesCollection()
        
        let pricesPreTax = WBPreferences.enteredPricePreTax()
        let reportOnlyReimbursable = WBPreferences.onlyIncludeReimbursableReceiptsInReports()
    
        var hasNonReimbursable = false
        
        let recs = receipts()
        
        for receipt in recs {
            if reportOnlyReimbursable && !receipt.isReimbursable {
                continue
            }
            let price = receipt.targetPrice()
            let tax = receipt.targetTax()
            
            receiptTotal.addPrice(price)
            taxesTotal.addPrice(tax)
            noTaxesTotal.addPrice(price)
            grandTotal.addPrice(receipt.price())
            
            if WBPreferences.enteredPricePreTax() {
                grandTotal.addPrice(tax)
                receiptTotal.addPrice(tax)
            } else {
                noTaxesTotal.subtractPrice(tax)
            }
            
            if receipt.isReimbursable {
                reimbursableTotal.addPrice(price)
                if WBPreferences.enteredPricePreTax() {
                    reimbursableTotal.addPrice(tax)
                }
            } else {
                hasNonReimbursable = true
            }
        }
        
        let dists = distances()
        
        for distance in dists {
            grandTotal.addPrice(distance.totalRate())
            reimbursableTotal.addPrice(distance.totalRate())
            distanceTotal.addPrice(distance.totalRate())
        }
        
        pdfRender.setTripName(trip: trip.name)
        
        let startDate = dateFormatter.formattedDate(trip.startDate, in: trip.startTimeZone)!
        let endDate = dateFormatter.formattedDate(trip.endDate, in: trip.endTimeZone)!
        pdfRender.appendHeader(row: String(format: WBPreferences.loclized(key: "report_header_duration"), startDate, endDate))
        
        if let costCenter = trip.costCenter, WBPreferences.trackCostCenter() && !costCenter.isEmpty {
            pdfRender.appendHeader(row: String(format: WBPreferences.loclized(key: "report_header_comment"), costCenter))
        }
        
        if let comment = trip.comment, !trip.comment.isEmpty {
            pdfRender.appendHeader(row: String(format: WBPreferences.loclized(key: "report_header_comment"), trip.comment))
        }
        
        // Short Code for Get: currencyFormattedPrice()
        func fp(_ p: PricesCollection) -> String { return p.currencyFormattedPrice() }
        
        
        // Print the various tax totals if the IncludeTaxField is true and we have taxes
        if WBPreferences.includeTaxField() && taxesTotal.hasValue() {
            pdfRender.appendHeader(row: String(format: WBPreferences.loclized(key: "report_header_receipts_total_no_tax"), fp(noTaxesTotal)))
            pdfRender.appendHeader(row: String(format: WBPreferences.loclized(key: "report_header_receipts_total_tax"), fp(taxesTotal)))
            pdfRender.appendHeader(row: String(format: WBPreferences.loclized(key: "report_header_receipts_total_with_tax"), fp(receiptTotal)))
        } else if dists.count > 0 {
            // Prints the receipts total if we have distances AND (the IncludeTaxField setting is false OR the value of taxes is 0)
            // We use this to distinguish receipts vs distances when we do NOT have the tax breakdown
            pdfRender.appendHeader(row: String(format: WBPreferences.loclized(key: "report_header_receipts_total"), fp(receiptTotal)))
        }
        
        // Print out the distances (if any)
        if dists.count > 0 {
            pdfRender.appendHeader(row: String(format: WBPreferences.loclized(key: "report_header_distance_total"), fp(distanceTotal)))
        }
        
        // Print the grand total
        pdfRender.appendHeader(row: String(format: WBPreferences.loclized(key: "report_header_grand_total"), fp(grandTotal)), style: .defaultBold)
        
        // Print the grand total (reimbursable)
        if !reportOnlyReimbursable && hasNonReimbursable {
            let row = String(format: WBPreferences.loclized(key: "report_header_receipts_total_reimbursable"), fp(reimbursableTotal))
            pdfRender.appendHeader(row: row, style: .defaultBold)
        }
        
        pdfRender.closeHeader()
        
        
        //MARK: Render Tables
        
        if !WBPreferences.omitDefaultPdfTable() || !PurchaseService.hasValidSubscriptionValue  {
            appendReportsTable()
        }

        if WBPreferences.printDistanceTable() && !dists.isEmpty {
            appendDistancesTable()
        }

        if WBPreferences.includeCategoricalSummation() && PurchaseService.hasValidSubscriptionValue {
            appendCategoricalSummationTable()
        }
        
        if WBPreferences.separatePaymantsByCategory() && PurchaseService.hasValidSubscriptionValue {
            appendSeparatedByCategoryTable()
        }
        
    }
    
    private func appendSeparatedByCategoryTable() {
        for (category, receipts) in receiptsByCategories() {
            let receiptsTable = ReportPDFTable(title: category, pdfRender: pdfRender, columns: receiptColumns())!
            receiptsTable.includeHeaders = true
            receiptsTable.includeFooters = true
            receiptsTable.append(withRows: receipts)
        }
    }
    
    private func appendReportsTable() {
        var rows = receipts()
        let receiptsTable = ReportPDFTable(pdfRender: pdfRender, columns: receiptColumns())!
        receiptsTable.includeHeaders = true
        receiptsTable.includeFooters = true
        if WBPreferences.printDailyDistanceValues() {
            let dReceipts = DistancesToReceiptsConverter.convertDistances(distances()) as! [WBReceipt]
            rows.append(contentsOf: dReceipts)
            rows.sort(by: { $1.date.compare($0.date) == .orderedAscending })
        }
        receiptsTable.append(withRows: rows)
    }
    
    private func appendDistancesTable() {
        let distancesTable = ReportPDFTable(pdfRender: pdfRender, columns: distanceColumns())!
        distancesTable.includeHeaders = true
        distancesTable.includeFooters = true
        distancesTable.append(withRows: distances())
    }
    
    private func appendCategoricalSummationTable() {
        let receipts = Array(receiptsByCategories().values)
        
        let isDefaultCurrencies = receipts
            .flatMap { $0 }
            .filter { $0.currency != trip.defaultCurrency }
            .isEmpty
        
        var columns = categoryColumns()
        if isDefaultCurrencies { columns.removeLast() }
        
        let categoricalTable = ReportPDFTable(pdfRender: pdfRender, columns: columns)!
        categoricalTable.includeHeaders = true
        categoricalTable.includeFooters = false
        categoricalTable.append(withRows: receipts)
    }
    
    
}

